src/mlpack/methods/reinforcement_learning/replay/hindsight_replay.hpp

```diff
@@ -0,0 +1,423 @@
+ * from failed attempts. It allows sample-efficient learning in 
+ * sparse and binary reward situations.
+ *
+ * @code
+ * @article{marcin2017hindsight,
+ *  title   = {Hindsight Experience Replay},
+ *  author  = {Marcin Andrychowicz, Filip Wolski, Alex Ray,
+ *             Jonas Schneider, Rachel Fong, Peter Welinder, Bob McGrew,
+ *             Josh Tobin, Pieter Abbeel, Wojciech Zaremba},
+ *  journal = {arXiv preprint arXiv:1707.01495},
+ *  year    = {2017}
+ *  }
+ * @endcode
+ *
+ * @tparam EnvironmentType Desired task.
+ */
+template <typename EnvironmentType>
+class HindsightReplay
+{
+ public:
+  //! Convenient typedef for action.
+  using ActionType = typename EnvironmentType::Action;
+
+  //! Convenient typedef for state.
+  using StateType = typename EnvironmentType::State;
+
+  struct Transition
+  {
+    StateType state;
+    ActionType action;
+    double reward;
+    StateType nextState;
+    bool isEnd;
+    StateType goal;
+  };
+
+  enum goalStrategy
+  {
+    FINAL,
+    FUTURE,
+    RANDOM,
+    EPISODE,
+  };
+
+  /**
+   * Default constructor.
+   */
+  HindsightReplay():
+      batchSize(0),
+      capacity(0),
+      position(0),
+      full(false),
+      herRatio(1),
+      goalSelectionStrategy(goalStrategy::FUTURE),
+      nSteps(0)
+  { /* Nothing to do here. */ }
+
+  /**
+   * Construct an instance of hindsight experience replay class.
+   *
+   * @param batchSize Number of examples returned at each sample.
+   * @param capacity Total memory size in terms of number of examples.
+   * @param herRatio ratio of HER  to data coming from normal 
+   * experience replay in replay buffer
+   * @param nSteps Number of steps to look in the future.
+   * @param dimension The dimension of an encoded state.
+   * @param strategy goal selection startegy for HER
+   */
+  HindsightReplay(const size_t batchSize,
+                    const size_t capacity,
+                    const size_t herRatio = 4,
+                    EnvironmentType environment = EnvironmentType(),
+                    goalStrategy strategy = goalStrategy::FUTURE,
+                    const size_t nSteps = 1,
+                    const size_t dimension = StateType::dimension) :
+      batchSize(batchSize),
+      capacity((herRatio+1)*capacity),
+      position(0),
+      full(false),
+      herRatio(herRatio),
+      goalSelectionStrategy(strategy),
+      nSteps(nSteps),
+      states(dimension, (herRatio+1)*capacity),
+      actions((herRatio+1)*capacity),
+      rewards((herRatio+1)*capacity),
+      nextStates(dimension, (herRatio+1)*capacity),
+      goals(dimension, (herRatio+1)*capacity),
+      isTerminal((herRatio+1)*capacity),
+      environment(environment)
+  { /* Nothing to do here. */ }
+
+  /**
+   * Sample goal according to goal strategy
+   * 
+   * @param transitionIndex index of current transition
+   */
+  void SampleGoal(StateType& desiredGoal, int transitionIndex)
+  {
+    if (goalSelectionStrategy == goalStrategy::FINAL)
```

Comment:
python doesn't have case switch doesn't mean we don't.

Reply:
Fair Point. I will make that change.